keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1239 @@
+Furthermore, if the logging container is not yet started and the main app
+crashes on startup, those logs can be lost (depends if logs go to a shared
+volume or over the network on localhost, etc.). While you can modify your
+application to handle this scenario during startup (as it is probably the change
+you need to do to handle sidecar crashes), for shutdown this approach won't
+work.
+
+On shutdown the ordering behavior is arguably more important: if the logging
+container is stopped first, logs for other containers are lost. No matter if
+those containers queue them and retry to send them to the logging container, or
+if they are persisted to a shared volume. The logging container is already
+killed and will not be restarted, as the pod is shutting down. In these cases,
+logs are lost.
+
+The same things regarding startup and shutdown apply for a metrics container.
+
+### Problems: service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will
+be blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic functionality.
+
+Some service meshes depend on secret (like a certificate) downloaded by other
+Init container to establish secure communication between services. This makes
+the problem of ordering of sidecar and Init containers harder.
+
+### Problems: configuration / secrets
+
+Some pods use init containers to pull down configuration/secrets and update them
+before the main container gets it. Then use sidecars to continue to watch for
+changes and perform the updates and push to the main container. This requires
+two separate code paths today. Perhaps the same sidecar could be used for both
+cases.
+
+### Goals
+
+This proposal aims to:
+- make containers implementing the sidecar pattern first class citizens inside a
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- allow to enforce security boundaries to sidecar containers different that
+  other containers. For example, allow Istio to run privileged to configure ip
+  tables and disable other containers from doing so
+- (alpha) support containers termination ordering
```

Comment:
This doesn't make sense to me. Either we do have support for ordering container termination, or we don't. If it's a goal but we don't mind that ab alpha doesn't include it yet, we should list it as a goal and then also mention the relevant details under Graduation Criteria.

Reply:
@jpbetz what do you think of stating the goal: "Allow to implement sidecar containers that will guaranteed to be running longer than regular containers"? I don't think ordering is something we are ready to commit on.

Reply:
I think we have a goal that sidecar containers are guaranteed to shutdown strictly after app containers.  I strongly suspect that this KEP will require more than that, but it's still a suspicion rather than a goal.  It's not a non-goal.

I'd call it out under goals as "to be evaluated and decided before beta"

Reply:
Agree that we should include the goal of shutting sidecars after apps :+1:

Reply:
added